### PR TITLE
[SPARK-25222][K8S] Improve container status logging

### PR DESCRIPTION
**Pull request by rvesse**
## What changes were proposed in this pull request?

Currently when running Spark on Kubernetes a logger is run by the client that watches the K8S API for events related to the Driver pod and logs them.  However for the container status aspect of the logging this simply dumps the raw object which is not human readable e.g.

screen shot 2018-08-24 at 10 37 46
screen shot 2018-08-24 at 10 38 14

This is despite the fact that the logging class in question actually has methods to pretty print this information but only invokes these at the end of a job.

This PR improves the logging to always use the pretty printing methods, additionally modifying them to include further useful information provided by the K8S API.

A similar issue also exists when tasks are lost that will be addressed by further commits to this PR

- [x] Improved `LoggingPodStatusWatcher`
- [x] Improved container status on task failure

## How was this patch tested?

Built and launched jobs with the updated Spark client and observed the new human readable output:

screen shot 2018-08-24 at 11 09 32
screen shot 2018-08-24 at 11 09 42
screen shot 2018-08-24 at 11 10 13
screen shot 2018-08-24 at 17 47 44

Suggested reviewers: @liyinan926 @mccheah

### CODE REVIEW

**Review by liyinan926 — APPROVED** [review]

**Comment by liyinan926** [comment]
@mccheah can you give `ok to test` to this one and help merge it?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala [hunk 1]
```diff
@@ -60,4 +64,81 @@ private[spark] object KubernetesUtils {
   }
 
   def parseMasterUrl(url: String): String = url.substring("k8s://".length)
+
+  def formatPairsBundle(pairs: Seq[(String, String)], indent: Int = 1) : String = {
+    // Use more loggable format if value is null or empty
+    val indentStr = "\t" * indent
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: Can we prefer space-based indentation? Curious as to whether others have an opinion about this.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala [hunk 2]
```diff
@@ -60,4 +64,81 @@ private[spark] object KubernetesUtils {
...
+        ("container name", status.getName),
+        ("container image", status.getImage)) ++
+        containerStatusDescription(status)
+    }.map(p => formatPairsBundle(p, indent)).mkString("\n\n")
+  }
+
+  def containerStatusDescription(containerStatus: ContainerStatus)
+    : Seq[(String, String)] = {
+    val state = containerStatus.getState
+    Option(state.getRunning)
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: This is a really cool use of partial functions - wonder if there's other places where we should be matching this way (of course doesn't have to be done here).

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsLifecycleManager.scala [hunk 3]
```diff
@@ -151,13 +152,15 @@ private[spark] class ExecutorPodsLifecycleManager(
 
   private def exitReasonMessage(podState: FinalPodState, execId: Long, exitCode: Int) = {
     val pod = podState.pod
+    val reason = Option(pod.getStatus.getReason)
+    val message = Option(pod.getStatus.getMessage)
     s"""
        |The executor with id $execId exited with exit code $exitCode.
-       |The API gave the following brief reason: ${pod.getStatus.getReason}
-       |The API gave the following message: ${pod.getStatus.getMessage}
+       |The API gave the following brief reason: ${reason.getOrElse("")}
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsLifecycleManager.scala]: Maybe default as `N/A`? Might be confusing to be left blank.

**Comment by mccheah** [comment]
ok to test

**Review by mccheah — CHANGES_REQUESTED** [review]
Definitely looks a lot prettier - some small comments but otherwise looks good.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala (hunk 1 above)
Comment (rvesse) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: I just preserved the original codes choice here, I would happily change to spaces if preferred

**Comment by rvesse** [comment]
@mccheah Thanks for the review, have made the change you suggested to use N/A instead of empty string.

I have left indentation as tabs for now, as I said in a previous comment this was just what the existing code used and I am happy to change it if others also want the change to spaces made

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsLifecycleManager.scala (hunk 3 above)
Comment (rvesse) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsLifecycleManager.scala]: Fixed in latest commit

**Comment by rvesse** [comment]
Think this is pretty much ready to merge, can folks take another look when they get chance

**Comment by mccheah** [comment]
Yup this can merge now, thanks!

**Pull request closed without merging**